### DISCUSSION TITLE
Internal Type Inference

### DISCUSSION THREAD

**Discussion by Kludex** [astral-sh/ruff#5686] [Q&A] [answered]
Hi @charliermarsh 👋 

I'm building [bump-pydantic](https://github.com/pydantic/bump-pydantic), and I was wondering if you can help me.

You don't do any kind of type inference on `ruff`, right?

I want to refactor the following code:

```py
from pydantic import BaseModel

class Model(BaseModel):
    a: int

model = Model(a=1)
model.dict()  # change this to `model.model_dump()`
```

I want to make sure that `model` is a `BaseModel` to perform the refactor.

Do you do something like that in `ruff`?

**Comment by charliermarsh** [comment, marked as answer]
Yeah, we don't do this in Ruff right now. Our semantic model has enough information that we _could_ support a rewrite in some cases like the one above, but there's a sliding scale of complexity, e.g. this is harder:

```python
def f(x: int) -> Any:
  return Model(a=1)

model = f()
model.dict()
```

This is also harder:

```python
from other_module import Model

model = Model(a=1)
model.dict()
```

**Reply by Kludex** [reply]
Yeah, I thought so.

Thanks anyway. 🙏

**Reply by charliermarsh** [reply]
You could look at using [astroid](https://pypi.org/project/astroid/) which has some support for static inference.
